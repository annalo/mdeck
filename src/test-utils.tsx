/* eslint-disable import/no-extraneous-dependencies */

import React from "react";
import { render } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";

import { ThemeProvider } from "components/Home/ThemeProvider";
import { MarkdownContextProvider } from "contexts/MarkdownContext";
import { SlideObserverProvider } from "contexts/SlideObserver";
import { CodeLineObserverProvider } from "contexts/CodeLineObserver";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider>
      <MarkdownContextProvider>
        <SlideObserverProvider>
          <CodeLineObserverProvider>{children}</CodeLineObserverProvider>
        </SlideObserverProvider>
      </MarkdownContextProvider>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

// override render and renderHook methods
export { customRender as render, customRenderHook as renderHook };
