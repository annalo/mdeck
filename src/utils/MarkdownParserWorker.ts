import Worker from "worker-loader!./Worker"; // eslint-disable-line

import type { Dispatch } from "react";
import type { MarkdownContextReducerAction } from "types/markdown-context";

import { MarkdownContextReducerActionType } from "types/markdown-context";

export default class MarkdownParserWorker {
  private readonly worker: Worker;

  constructor(dispatch: Dispatch<MarkdownContextReducerAction>) {
    this.worker = new Worker();
    this.worker.onmessage = ({ data }: { data: HtmlArray }) =>
      dispatch({
        type: MarkdownContextReducerActionType.SetHtmlArray,
        htmlArray: data,
      });
  }

  parse(md: MarkdownString): void {
    this.worker.postMessage(md);
  }

  terminate(): void {
    if (this.worker) this.worker.terminate();
  }
}
