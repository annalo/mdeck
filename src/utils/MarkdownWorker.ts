import Worker from "worker-loader!./Worker"; // eslint-disable-line
import type { Dispatch } from "react";
import type { MarkdownContextReducerAction } from "types/markdown-context";

export class MarkdownWorker {
  private readonly worker: Worker;

  constructor(dispatch: Dispatch<MarkdownContextReducerAction>) {
    this.worker = new Worker();
    this.worker.onmessage = ({
      data,
    }: {
      data: MarkdownContextReducerAction;
    }) => {
      console.log("markdown worker on message", data);
      dispatch(data);
    };
  }

  parse(md: string): void {
    this.worker.postMessage({ action: "PARSE", md });
  }

  terminate() {
    if (this.worker) this.worker.terminate();
  }
}
