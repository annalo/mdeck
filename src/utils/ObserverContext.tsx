import { createContext } from "react";

export const ObserverContext = createContext<IntersectionObserver | null>(null);
