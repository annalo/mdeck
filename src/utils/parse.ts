import { Marpit } from "@marp-team/marpit";
import TaskList from "markdown-it-task-lists";
import { injectLineNumber } from "utils/parsePlugins/injectLineNumber";

export const parse = (markdown: string): HtmlArray => {
  const marpit = new Marpit({
    inlineSVG: true,
    markdown: { html: true, linkify: true, typographer: true },
  })
    .use(TaskList)
    .use(injectLineNumber);

  const { html }: { html: HtmlArray } = marpit.render(markdown, {
    htmlAsArray: true,
  });
  return html;
};
