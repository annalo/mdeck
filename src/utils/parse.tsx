/* eslint-disable */
import { Marpit } from "@marp-team/marpit";
import TaskList from "markdown-it-task-lists";

const RULES = [
  "blockquote_open",
  "code_block",
  "fence",
  "heading_open",
  "image",
  "link_open",
  "list_item_open",
  "paragraph_open",
  "table_open",
];

function injectLineNumber(md: any) {
  const { marpit_inline_svg_open } = md.renderer.rules;

  // Enable line sync by per slides
  md.renderer.rules.marpit_inline_svg_open = (tokens, idx, opts, env, self) => {
    const slide = tokens
      .slice(idx + 1)
      .find((t) => t.type === "marpit_slide_open");

    if (slide.map?.length) {
      tokens[idx].attrJoin("class", "src-line");
      tokens[idx].attrSet("data-line", slide.map[0]);
    }

    const renderer = marpit_inline_svg_open || self.renderToken;
    return renderer.call(self, tokens, idx, opts, env, self);
  };

  // Enables line sync per elements
  for (const rule of RULES) {
    const original = md.renderer.rules[rule];

    md.renderer.rules[rule] = (tokens, idx, options, env, self) => {
      const token = tokens[idx];

      if (token.map?.length) {
        token.attrJoin("class", "code-line");
        token.attrSet("data-line", token.map[0]);
      }

      const renderer = original || self.renderToken;
      return renderer.call(self, tokens, idx, options, env, self);
    };
  }
}

export function parse(markdown: string): string {
  const marpit = new Marpit({
    inlineSVG: true,
    markdown: { html: true, linkify: true, typographer: true },
  })
    .use(TaskList)
    .use(injectLineNumber);

  const { html }: { html: string } = marpit.render(markdown);
  return html;
}
