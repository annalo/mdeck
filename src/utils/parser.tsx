import { Marpit } from "@marp-team/marpit";

export function convertMarkdown(markdown: string): string {
  const marpit = new Marpit();
  const { html } = marpit.render(markdown);
  return html;
}
