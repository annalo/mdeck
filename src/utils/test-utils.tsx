/* eslint-disable import/no-extraneous-dependencies */
import React from "react";
import { render } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";

import { ThemeProvider } from "components/App/ThemeProvider";

import { MarkdownContextProvider } from "contexts/MarkdownContext";
import { PresentationContextProvider } from "contexts/PresentationContext";
import { SlideObserverProvider } from "contexts/SlideObserver";
import { CodeLineObserverProvider } from "contexts/CodeLineObserver";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider>
      <MarkdownContextProvider>
        <PresentationContextProvider>
          <SlideObserverProvider>
            <CodeLineObserverProvider>{children}</CodeLineObserverProvider>
          </SlideObserverProvider>
        </PresentationContextProvider>
      </MarkdownContextProvider>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

// override render and renderHook methods
export { customRender as render, customRenderHook as renderHook };
