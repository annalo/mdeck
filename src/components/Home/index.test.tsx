import React from "react";
import { render } from "utils/test-utils";
import { Home } from ".";

describe("<Home />", () => {
  test("should render and match the snapshot", () => {
    const { asFragment } = render(<Home />);
    expect(asFragment()).toMatchSnapshot();
  });
});
