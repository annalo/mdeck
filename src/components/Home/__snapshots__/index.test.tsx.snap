// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render and match the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  height: 100%;
}

<div
    class="c0"
    id="main"
  />
</DocumentFragment>
`;
